hugr-model/src/v0/mod.rs

```diff
@@ -0,0 +1,368 @@
+    use super::Operation;
+    use super::{Scheme, Symbol, TermId};
+
+    /// See [`Operation::DefineFunc`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct DefineFunc {
+        /// The name of the function to be defined.
+        pub name: Symbol,
+        /// The type scheme of the function.
+        pub r#type: Box<Scheme>,
+    }
+
+    /// See [`Operation::DeclareFunc`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct DeclareFunc {
+        /// The name of the function to be declared.
+        pub name: Symbol,
+        /// The type scheme of the function.
+        pub r#type: Box<Scheme>,
+    }
+
+    /// See [`Operation::CallFunc`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct CallFunc {
+        /// The name of the function to be called.
+        pub name: Symbol,
+    }
+
+    /// See [`Operation::LoadFunc`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct LoadFunc {
+        /// The name of the function to be loaded.
+        pub name: Symbol,
+    }
+
+    /// See [`Operation::Custom`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct Custom {
+        /// The name of the custom operation.
+        pub name: Symbol,
+    }
+
+    /// See [`Operation::DefineAlias`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct DefineAlias {
+        /// The name of the alias to be defined.
+        pub name: Symbol,
+        /// The value of the alias.
+        pub value: TermId,
+    }
+
+    /// See [`Operation::DeclareAlias`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct DeclareAlias {
+        /// The name of the alias to be declared.
+        pub name: Symbol,
+        /// The type of the alias.
+        pub r#type: TermId,
+    }
+}
+
+/// A metadata item.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct MetaItem {
+    /// Name of the metadata item.
+    pub name: SmolStr,
+    /// Value of the metadata item.
+    pub value: Term,
+}
+
+/// A port in the graph.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Port {
+    /// The id of the edge that the port is connected to.
+    /// All ports referencing the same edge are connected.
+    pub edge: EdgeId,
+
+    /// The type of the port.
+    ///
+    /// This must be a term of type `Type`.
+    /// If the type is unknown, this will be a wildcard term.
+    pub r#type: TermId,
+
+    /// Metadata attached to the port.
+    pub meta: Vec<MetaItem>,
+}
+
+/// Schemes are parameterized terms.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Scheme {
+    /// The named parameters of the scheme.
+    ///
+    /// Within any term, the previous terms of the parameter list are available as variables.
+    pub params: Vec<SchemeParam>,
+    /// Constraints on the parameters of the scheme.
+    ///
+    /// All parameters are available as variables within the constraints.
+    /// The constraints must be terms of type `Constraint`.
```

Comment:
Right - I'm puzzled here. I don't see any `enum Term` variants that have type `Constraint`; I see only one for the type *of* Constraints, and I can't understand how that would be used. Am I missing something, looking at this the wrong way, or is this just WIP?

Reply:
This is all a big WIP.